feat: 푸시 알림

duol9 opened the pull request:
## 🌱 관련 이슈
- close #61 

## 📌 작업 내용 및 특이사항
- FCM 연동 설정
- 트랜잭션 성공 후 알림 발송을 위해 이벤트 리스너 적용
- FCM 토큰 저장을 위해 NotificationToken 도메인 추가
- 상태 변경, 룸메 참여(초대코드 입력)시 알림이 가도록 로직 추가

## 📝 참고사항
- 노트 생성에 대한 알림 전송로직은 추후 추가하도록 하겠습니다. 
- **보안정보에 FCM 시크릿키 내용을 추가했습니다.** 
- 이해를 돕기 위해 **커밋 본문에 부가 설명을 작성**했습니다.
- FCM 관련 환경변수가 주입되지 않아 빈 생성이 안되는 오류가 발생하였습니다. 현재는 임시대책으로 프로파일이 'dev'일 때만 빈이 생성되도록 어노테이션을 추가하였습니다. Config에 환경변수를 직접 주입하지 않는 방법이 도입된 후 다시 한 번 만져보도록 하겠습니다. (@Profile("!test")를 붙여야 하는 파일이 2개다 보니 관리 포인트가 늘어 불편하기도 하고 근본적인 해결책이라 생각되지 않습니다 ㅠㅠ)
- 룸메 상태 조회를 구현하면서 만든 findRoommateInMyRoom 메서드는 상태를 항상 join 하고, roomId를 알아야 하는 제약이 있었습니다. 하지만 실제 사용 패턴을 고려했을 때는
1. 룸메를 알고 있어도 roomId는 모르는 경우가 더 많고,
2. 룸메 상태까지 필요한 경우는 제한적이라 **불필요한 join이 발생**하고,
3. roomId를 반드시 알아야 하는 상황도 생각보다 적다는 점을 확인했습니다.

그래서 roomId 없이 memberId만으로 룸메를 조회할 수 있는 메서드(findRoommateByMemberId)를 추가했습니다.

추가적으로, 기존 메서드명을 findRoommateInMyRoom에서 findRoommateWithStatusByMemberId로 바꿔서 roomId 제약을 없애는 방향으로 리팩토링하는 것이 더 낫지 않을까 고민 중입니다. (Room을 찾는 메서드는 쓰임새의 가능성을 위해 생각해 남겨두긴 하겠지만, 생각보다 roomID가 필요한 상황이 적을 것 같습니다.. 또 상태 패치조인 없이 룸메 조회하는 메서드명과 일관성을 갖출 수도 있습니다!)

기존
<img width="1216" height="456" alt="image" src="https://github.com/user-attachments/assets/0de97650-7aaf-4070-aa52-fa5e16e49eb0" />
<img width="1344" height="502" alt="image" src="https://github.com/user-attachments/assets/bb8058d1-9c62-4dbd-a275-a5c7dd0aa508" />

변경한다면 
<img width="1130" height="770" alt="image" src="https://github.com/user-attachments/assets/750a8900-e262-4a1c-b6a3-9261c917ea0e" />


## 📚 기타
-


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신기능**
  - 룸메이트 합류 및 상태 변경 시 푸시 알림이 발송되는 이벤트 기반 알림 흐름이 도입되었습니다.
  - 디바이스 토큰 저장 API와 토큰 관리/메시지 변환/전송 파이프라인이 추가되어 푸시 발송을 지원합니다.
- **유지보수**
  - Firebase 초기화 및 연동 설정과 비동기 실행기(notificationExecutor)가 추가되었습니다.
  - 빌드 의존성 업데이트 및 Firebase 경로를 .gitignore에 추가했습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

duol9 committed:
chore: FCM 설정

duol9 committed:
chore: FCM 에러코드 추가

duol9 committed:
feat: Notification 엔티티 작성

duol9 committed:
feat: Notification 생성 dto 추가

duol9 committed:
feat: Notification 삭제 메서드 추가

duol9 committed:
feat: Notification 토큰 저장 서비스 로직 구현

duol9 committed:
feat: Notification 토큰 저장 API 추가

duol9 committed:
refactor: entity에서 domain으로 패키지명 변경

duol9 committed:
refactor: Notification 엔티티명 변경

- fcm 라이브에 포함되어 있는 클래스와 이름이 겹쳐 Notification -> NotificationToken 으로 변경

duol9 committed:
chore: 알림 비동기 처리 설정

- API 응답과 분리하기 위해 비동기 설정을 추가
- ec2 t2.micro 스펙에 맞춰 스레드를 최소한으로 설정

duol9 committed:
feat: 알림 타입 작성

duol9 committed:
feat: 알림 발송 서비스 인터페이스 정의

- 도메인 계층이 FCM 같은 외부 클라이언트 구현체에 직접 의존하지 않도록 알림 발송을 위한 NotificationSender 인터페이스를 정의

duol9 committed:
feat: 알림토큰 조회 메서드 추가

duol9 committed:
feat: FCM 메시지 dto 추가

duol9 committed:
feat: FCM 메시지 변환 메서드 추가

- FcmMessageRequest를 Firebase Admin SDK의 'Message' 객체로 변환하기 위해 작성
- createApnsConfig(): IOS 알림 설정, 화면이 꺼져있어도 알림 발송 허용 (setContentAvailable(true))
- createAndroidConfig(): Android 알림 설정, 중요도 보통으로 지정해 일반적인 알림으로 설정 (setPriority(Priority.NORMAL))

duol9 committed:
feat: FCM 메시지 전송 로직 구현

- 다중 폰 사용 등 회원이 여러 토큰을 가질 경우를 고려해 알림 토큰을 List로 조회
- FCM 서버로부터 'UNREGISTERED' 에러를(예: 사용자가 앱 삭제) 받으면 해당 토큰을 DB에서 삭제되도록 처리
- 알림 발송은 별도의 API가 아닌 부가 기능이므로 실패하더라도 클라이언트에서 취할 조치가 없다고 판단하여 서버 로그로만 남기도록 처리

duol9 committed:
feat: 이벤트 기반 비동기 알림 발송 기능 추가

- 핵심 로직(쪽지 생성, 상태 변경, 룸메 참가)의 트랜잭션이 성공적으로 커밋되었을 때만 알림 발송 이벤트가 실행하도록 @TransactionalEventListener 사용

duol9 committed:
feat: 룸메 조회 메서드 추가

- roomId가 없어도 룸메를 찾을 수 있도록 메서드 추가
- 룸메만 찾을 때는 기존 findRoommateInMyRoom에서 이루어지는 상태 조인이 불필요하다 판단하여 별도의 메서드로 생성

duol9 committed:
feat: 상태 변경, 룸메 참가 서비스 로직에 알림 발송 메서드 추가

duol9 committed:
refactor: Firebase 키 주입 방식을 환경 변수로 변경

- 기존의 파일 경로(`file-path`) 방식은 로컬(classpath:)과 서버(file:)의 경로를 다르게 설정했어야 했는데, 주입 방식을 환경변수로 통일해 동일한 환경이 되도록 변경

duol9 committed:
test: 테스트용 가짜 알림 발송 메서드 추가

duol9 committed:
fix: 테스트 환경에서 FCM 관련 Bean 제외

- 테스트 실행 시 FCM 환경변수 누락으로 인한 오류 해결

ohhamma commented on src/main/java/com/example/wini/infra/fcm/client/FcmClient.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private void handleFirebaseMessagingException(FirebaseMessagingException e, NotificationToken notificationToken) {
+        MessagingErrorCode errorCode = e.getMessagingErrorCode();
+        if (errorCode == MessagingErrorCode.UNREGISTERED) {
+            notificationTokenRepository.delete(notificationToken);
+        } else {
+            log.error(
+                    "알림 발송 실패. token: {}, errorCode: {}, error: {}",
+                    notificationToken.getToken(),
+                    errorCode,
+                    e.getMessage());
+        }
+    }
```
FirebaseMessagingException을 GlobalExceptionHandler에서 처리하는것도 혹시 가능할까요?

duol9 replied:
handleFirebaseMessagingException 이 메서드가 하는 일을 GlobalExceptionHandler가 하도록 하는걸 말씀하시는 걸까요??

ohhamma replied:
네 맞습니다!

duol9 replied:
오 함 시도해볼게요!!

ohhamma commented on src/main/java/com/example/wini/infra/fcm/client/FcmClient.java:
```diff
@@ -0,0 +1,63 @@
+        List<NotificationToken> notificationTokens = notificationTokenRepository.findAllByMember_Id(recipientId);
+
+        if (notificationTokens.isEmpty()) {
+            log.info("알림 발송 대상 없음 - memberId: {}", recipientId);
```
서버에서 로그 남기는 형식에 대해서 이야기해보면 좋을 것 같습니다!
맨 앞에 로그를 발생시킨 위치를 같이 언급해주는 건 어떤가요?
```
[FcmClient] 알림 발송 대상 없음 - memberId: {}
```

duol9 replied:
좋은거 같아요!! 안그래도 로그 어떤식으로 남기면 좋을지 얘기해보려고 헸는데 그새 휘발됏네요

ohhamma replied:
좋습니다 :)

ohhamma reviewed (commented):
구현해주신 부분 확인했습니다!
어려운 코드가 정말 많네요.. 수고많으셨습니다.
저도 잘 모르는 구현 내용이 꽤 있어 이것저것 찾아보면서 리뷰를 진행했네요.

코드를 쭉 보면서 몇가지 궁금한 점에 대해서 코멘트로 남겨두었습니다.

1. `NotificationToken`은 `Notification`과 어떻게 다른건가요?
2. `NotificationToken`은 DB에 저장되어야 하는 값인가요?

추가로 `roomId` 없이 룸메를 조회할 수 있는 형식으로 함수를 변경한건 좋은 것 같습니다.

duol9 commented:
> 구현해주신 부분 확인했습니다! 어려운 코드가 정말 많네요.. 수고많으셨습니다. 저도 잘 모르는 구현 내용이 꽤 있어 이것저것 찾아보면서 리뷰를 진행했네요.
> 
> 코드를 쭉 보면서 몇가지 궁금한 점에 대해서 코멘트로 남겨두었습니다.
> 
> 1. `NotificationToken`은 `Notification`과 어떻게 다른건가요?
> 2. `NotificationToken`은 DB에 저장되어야 하는 값인가요?
> 
> 추가로 `roomId` 없이 룸메를 조회할 수 있는 형식으로 함수를 변경한건 좋은 것 같습니다.

1. 변경 전 Notification을 말씀하시는 걸까요? 의미는 동일합니다. 다만 Firebase 라이브러리에서 제공하는 Notification(알림 내용 정의) 클래스와 클래스명이 겹쳐 확실하게 구분하는게 나을거 같아 엔티티 명을 NotificationToken으로 변경했습니다.  그리고 Notification이라고 하니 알림 내용도 저장이 되어야 할 것 같은 느낌도 들더라구요,... 사실 하는건 토큰 저장이 끝인데 😅
2. 필수적이진 않지만 서버에 저장하는게 효율적이기에 DB에 저장하는 방식을 택했습니다.  만일 저장하지 않는다면 클라이언트가 알림이 필요한 기능 API을 요청할 때마다 알림 토큰을 같이 전송해줘야 하는데 그렇게되면 너무 클라이언트에게 의존적인 상황이 발생할 것 같아 효율적인 방법이라 생각하지 않았습니다.

duol9 committed:
refactor: NotificationToken 엔티티명 변경

- 명확성을 위해 NotificationToken에서 FirebaseToken으로 변경

duol9 committed:
refactor: FirebaseMessagingException 로그 처리를 GlobalExceptionHandler로 이동

duol9 committed:
refactor: 컨벤션에 맞게 로그 문구 변경

duol9 committed:
feat: memberId로 룸메 상태를 조회하는 메서드 생성

- findRoommateWithStatusByMemberId 메서드가 추가됨에 따라 findRoommateInMyRoom를 삭제
- 룸메 상태 조회 서비스 로직도 불필요한 Room 조회 로직 삭제 및 findRoommateWithStatusByMemberId 메서드 반영

ohhamma reviewed (approved):
수정해주신 부분까지 모두 확인했습니다!
복잡한 로직이었는데 잘 마무리된 것 같아 다행입니다 😊
